src/zenml/logging/step_logging.py

```diff
@@ -0,0 +1,230 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""ZenML logging handler."""
+
+import io
+import logging
+import os
+import time
+from io import StringIO
+from logging import LogRecord
+from logging.handlers import TimedRotatingFileHandler
+from typing import Optional
+from uuid import uuid4
+
+from zenml.artifact_stores import BaseArtifactStore
+from zenml.io import fileio
+from zenml.logger import get_logger
+from zenml.logging import (
+    LOGS_HANDLER_INTERVAL_SECONDS,
+    LOGS_HANDLER_MAX_MESSAGES,
+    STEP_LOGGER_NAME,
+)
+from zenml.utils.io_utils import is_remote
+
+# Get the logger
+logger = get_logger(__name__)
+
+
+def prepare_logs_uri(
+    artifact_store: "BaseArtifactStore",
+    step_name: str,
+    log_key: Optional[str] = None,
+) -> str:
+    """Generates and prepares a URI for the log file for a step.
+
+    Args:
+        artifact_store: The artifact store on which the artifact will be stored.
+        step_name: Name of the step.
+        log_key: The unique identification key of the log file.
+
+    Returns:
+        The URI of the logs file.
+    """
+    if log_key is None:
+        log_key = str(uuid4())
+
+    logs_base_uri = os.path.join(
+        artifact_store.path,
+        step_name,
+        "logs",
+    )
+
+    # Create the dir
+    if not fileio.exists(logs_base_uri):
+        fileio.makedirs(logs_base_uri)
+
+    # Delete the file if it already exists
+    logs_uri = os.path.join(logs_base_uri, f"{log_key}.log")
+    if fileio.exists(logs_uri):
+        logger.warning(
+            f"Logs file {logs_uri} already exists! Removing old log file..."
+        )
+        fileio.remove(logs_uri)
+    return logs_uri
+
+
+class StepStdOut(StringIO):
+    """A replacement for the sys.stdout to turn outputs into logging entries.
+
+    When used in combination with the ZenHandler, this class allows us to
+    capture any print statements or third party outputs as logs and store them
+    in the artifact store.
+
+    Right now, this is only used during the execution of a ZenML step.
+    """
+
+    stdout_logger = logging.getLogger(STEP_LOGGER_NAME)
+
+    def write(self, message: str) -> int:
+        """Write the incoming string as an info log entry.
+
+        Args:
+            message: the incoming message string,
+        """
+        if message != "\n":
+            self.stdout_logger.info(message)
+        return len(message)
```

Comment:
shouldn't this also `print` the message ? otherwise all stdout messages will be suppressed from the actual stdout if artifact-store logging is enabled